porter-finance/v1-core
# Use Custom Errors Over Require

This PR changes our usage of require for custom errors. I have included a usage with a custom chai matcher to check the arguments.

## Why Custom Errors?
Custom errors were introduced in [solidity 0.8.4](https://blog.soliditylang.org/2021/04/21/custom-errors/). These are more gas efficient than require in most usage. The reason is because `require`, to give any information, need to use a `string`. Every character in the string increases gas use on both deploy and on every time a function is reverted. 

Custom errors solve this problem by using an actual function-like object that is included in the ABI and can be compiled into a function signature. This means that we can have a very long name such as `InvalidCollateralProvidedToBroker` and it'll compile into 2 bytes or something. In addition to this, we can pass back arguments upon revert. These arguments increase gas cost on both deploy and on subsequent calls, so these should be used sparingly. It's nice to have the option if we need it. 

## How can they be tested?
These custom errors can be tested the same way require is tested, but we can also test the arguments returned.

If not checking the message: 
```js
await expect(token.transfer(walletTo.address, 1007)).to.be.reverted;
```

If checking the message:
```js
await expect(token.transfer(walletTo.address, 1007))
  .to.be.revertedWith('Insufficient funds');
```

Note, the [matcher](https://github.com/EthWorks/Waffle/blob/master/waffle-chai/src/matchers/revertedWith.ts#L31) 
does `message.includes('revert') && message.includes(revertReason)`, so if your custom error was, for example, `InvalidSomething(address someAddress);`

The possible ways to test would be:

```js
await expect(contract.revertingMethod()).to.be.reverted;
await expect(contract.revertingMethod()).to.be.revertedWith(
  "InvalidSomething"
);
```

But if your error contains a string argument, you need to escape the strings like so:
```js
await expect(contract.revertingMethod()).to.be.revertedWith(
  "InvalidSomething(\"0x0000000000000000000\")"
);
```

NOTE: BigNumber is harder to test if it overflows JS number because the error reverts with a non-string version of the number, but we cannot get that accurately. Thought still needs put into this, or avoid checking the revert of large numbers.

I have created a custom matcher that is not super robust but works... 
```ts
declare global {
  export namespace Chai {
    // eslint-disable-next-line no-unused-vars
    interface Assertion {
      revertedWithArgs(errorName: string, ...args: any): Promise<void>;
    }
  }
}
export async function useCustomErrorMatcher() {
  use(function (chai) {
    chai.Assertion.addMethod("revertedWithArgs", function (errorName, ...args) {
      const expected = `${errorName}(${args
        .map((arg) => JSON.stringify(arg))
        .join(", ")})`;
      new chai.Assertion(this._obj).to.be.revertedWith(expected);
    });
  });
}
```


usage looks like:
```ts
useCustomErrorMatcher(); // need to call this before use (is there a better way to do this?)
await expect(
  contract.revertingMethod()
).to.be.revertedWithArgs(
  "ErrorName",
  "ErrorStringArg",
  123
);
```
